packages/content-management/favorites/favorites_server/src/favorites_service.ts

```diff
@@ -0,0 +1,129 @@
+
+    const favoriteIds = favoritesSavedObject?.attributes?.favoriteIds ?? [];
+
+    return { favoriteIds };
+  }
+
+  public async addFavorite({ id }: { id: string }): Promise<{ favoriteIds: string[] }> {
+    let favoritesSavedObject = await this.getFavoritesSavedObject();
+
+    if (!favoritesSavedObject) {
+      favoritesSavedObject = await this.deps.savedObjectClient.create(
+        favoritesSavedObjectType.name,
+        {
+          userId: this.userId,
+          type: this.type,
+          favoriteIds: [id],
+        },
+        {
+          id: this.getFavoriteSavedObjectId(),
+        }
+      );
+
+      return { favoriteIds: favoritesSavedObject.attributes.favoriteIds };
+    } else {
+      const newFavoriteIds = [
+        ...(favoritesSavedObject.attributes.favoriteIds ?? []).filter(
+          (favoriteId) => favoriteId !== id
+        ),
+        id,
+      ];
+
+      await this.deps.savedObjectClient.update(
+        favoritesSavedObjectType.name,
+        favoritesSavedObject.id,
+        {
+          favoriteIds: newFavoriteIds,
+        },
+        {
+          version: favoritesSavedObject.version,
+        }
+      );
+
+      return { favoriteIds: newFavoriteIds };
+    }
+  }
+
+  public async removeFavorite({ id }: { id: string }): Promise<{ favoriteIds: string[] }> {
+    const favoritesSavedObject = await this.getFavoritesSavedObject();
+
+    if (!favoritesSavedObject) {
+      return { favoriteIds: [] };
+    }
+
+    const newFavoriteIds = (favoritesSavedObject.attributes.favoriteIds ?? []).filter(
+      (favoriteId) => favoriteId !== id
+    );
+
+    await this.deps.savedObjectClient.update(
+      favoritesSavedObjectType.name,
+      favoritesSavedObject.id,
+      {
+        favoriteIds: newFavoriteIds,
+      },
+      {
+        version: favoritesSavedObject.version,
+      }
+    );
+
+    return {
+      favoriteIds: newFavoriteIds,
+    };
+  }
+
+  private async getFavoritesSavedObject(): Promise<SavedObject<FavoritesSavedObjectAttributes> | null> {
+    try {
+      const favoritesSavedObject =
+        await this.deps.savedObjectClient.get<FavoritesSavedObjectAttributes>(
+          favoritesSavedObjectType.name,
+          this.getFavoriteSavedObjectId()
+        );
+
+      return favoritesSavedObject;
+    } catch (e) {
+      if (SavedObjectsErrorHelpers.isNotFoundError(e)) {
+        return null;
+      }
+
+      throw e;
+    }
+  }
+
+  private getFavoriteSavedObjectId() {
+    if (!this.userId) {
+      // This should never happen, but it's better to throw an error than to create a saved object with an invalid ID
+      throw new Error('userId is required to create a favorite saved object');
+    }
+
+    if (!this.type) {
+      // This should never happen, but it's better to throw an error than to create a saved object with an invalid ID
+      throw new Error('type is required to create a favorite saved object');
+    }
```

Comment:
__question:__ cannot you just throw directly in the constructor if either `userId` or `type` is invalid (and maybe initialize ID as a `readonly` class member there as well to do it once per instance)?